#! /bin/sh /usr/share/dpatch/dpatch-run
## 091125_gc_check.dpatch by Bryan McLellan  <btm@loftninjas.org>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Expand GC check per upstream issue #2326

@DPATCH@
diff -urNad ruby1.8-1.8.7.174~/gc.c ruby1.8-1.8.7.174/gc.c
--- ruby1.8-1.8.7.174~/gc.c     2009-03-27 10:25:23.000000000 +0000
+++ ruby1.8-1.8.7.174/gc.c      2009-11-25 10:35:42.000000000 +0000
@@ -433,7 +433,7 @@
     if (during_gc)
        rb_bug("object allocation during garbage collection phase");

-    if (!malloc_limit || !freelist) garbage_collect();
+    if (!malloc_limit || !freelist || !freelist->as.free.next) garbage_collect();

     obj = (VALUE)freelist;
     freelist = freelist->as.free.next;
